### PR TITLE
expr: elide chains with one local get

### LINKED ISSUE #5076: Explain output contains redundancies
The `explain` output can contain noisy redundant expressions that don't really exist in the `RelationExpr`.

For example, TPCH Q04's explain output is
```
                          Optimized Plan                           
-------------------------------------------------------------------
 %0 =                                                             +
 | Get materialize.public.orders (u17)                            +
 | Filter (datetots(#4) < 1993-10-01 00:00:00), (#4 >= 1993-07-01)+
                                                                  +
 %1 =                                                             +
 | Get %0                                                         +
                                                                  +
 %2 =                                                             +
 | Get %0                                                         +
 | Filter !(isnull(#0))                                           +
 | Distinct group=(#0)                                            +
 | ArrangeBy (#0)                                                 +
                                                                  +
 %3 =                                                             +
 | Get materialize.public.lineitem (u7)                           +
 | Filter !(isnull(#0)), (#11 < #12)                              +
                                                                  +
 %4 =                                                             +
 | Join %2 %3 (= #0 #1)                                           +
 | | implementation = Differential %3 %2.(#0)                     +
 | | demand = (#0)                                                +
 | Distinct group=(#0)                                            +
 | ArrangeBy (#0)                                                 +
                                                                  +
 %5 =                                                             +
 | Join %1 %4 (= #0 #9)                                           +
 | | implementation = Differential %1 %4.(#0)                     +
 | | demand = (#5)                                                +
 | Reduce group=(#5)                                              +
 | | agg count(true)                                              +
                                                                  +
 Finish order_by=(#0 asc) limit=none offset=0 project=(#0, #1)    +
```
Up above, notice the
```
 %1 =                                                             +
 | Get %0                                                         +
```
which is clearly not very meaningful; each instance of `%1` could be replaced by `%0`.

What seems to be happening is that in `explain.rs`, we create a basic block for each input to a `Join` or `Union`, so that they can reference something when announcing their inputs. In some cases, those inputs may just be `Get { id: LocalId(_), .. }` and we could use the name associated with the let binding. Instead, we create a basic block containing just the `Get`, and use that name, as seen above.

I'm not really sure about the right "area" for this. The code is in the `expr` crate, but it is about user-facing stuff we document as "SQL".

### PR DESCRIPTION
**Pull request by benesch**
Clean up our EXPLAIN output by eliding chains that consist of only one
Get. Also annotate the chains that are bound to a proper Let, vs. those
that were introduced only to flatten the explain output.

Fix #5076.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5090)

### CODE REVIEW

File: test/sqllogictest/explain.slt [hunk 1]
```diff
@@ -118,29 +112,16 @@ EOF
...
-%3 =
-| Join %1 %2
+%2 =
+| Join %0 %1
 | | implementation = Unimplemented
-| | types = ()
-| | keys = ()
 | Map 1
-| | types = (i32)
-| | keys = ()
```
Comment (frankmcsherry) [inline test/sqllogictest/explain.slt:143]: Did something bad happen with these types and keys going away? I'm not super familiar so maybe intended?

Comment (benesch) [reply test/sqllogictest/explain.slt:143]: Oops!

Comment (benesch) [reply test/sqllogictest/explain.slt:143]: Ok, fixed! Good catch!

**Review by frankmcsherry — APPROVED** [review]
This looks great to me! I'm mostly looking at the output rather than the code, but I'm comfortable with that being shaken up substantially.

**Comment by benesch** [comment]
Whew, ok, I reworked the `sql` explanation code to match the `expr` explanation code, too. Stinks that they're duplicated but nearly identical, but don't see any good way to fix that.

**Pull request merged**